File: scala/pal/src/main/scala/pal/GalilRealtime.scala
package pal

import scala.collection.mutable

import ScaffoldGalil.{FPP_QUANTUM, OnlineGalil, OnlineSource}

/** Executable FIFO service specification; local SCA lowering is separate.
  *
  * The deque models the buffer for this reference. It is not claimed to be a
  * finite local transition table. Output decisions depend only on source events
  * and buffer emptiness; diagnostic counts do not guide the source or scheduler.
  *
  * Python 原典: `galil_realtime.py`。
  */
object GalilRealtime {

  /** A source behind a FIFO buffer that is granted `service` transitions per offered symbol. */
  class BufferedSource(val source: OnlineSource, val service: Int) {
    if (service < 1) {
      throw new IllegalArgumentException("positive integral source service required")
    }

    val pending: mutable.ArrayDeque[String] = mutable.ArrayDeque.empty

    // diagnostic counts
    var offered: Int = 0
    var reads: Int = 0
    var outputs: Int = 0
    var transitions: Int = 0
    var maxPending: Int = 0

    /** The last answer was the source's own output for the last offered symbol. */
    var lastCompleted: Boolean = false

    /** Offer one symbol, run the service, and answer for it (0 while still behind). */
    def read(char: String): Int = {
      if (char != "a" && char != "b") {
        throw new IllegalArgumentException("one binary input symbol required")
      }
      pending.append(char)
      offered += 1
      maxPending = math.max(maxPending, pending.size)
      var answer = 0
      lastCompleted = false
      var slot = 0
      var quiescent = false
      while (slot < service && !quiescent) {
        if (source.inputReady && pending.isEmpty) {
          quiescent = true // remaining service transitions are quiescent
        } else {
          val result = if (source.inputReady) {
            val readResult = source.read(pending.removeHead())
            reads += 1
            readResult
          } else {
            source.work()
          }
          transitions += 1
          result.output.foreach { value =>
            outputs += 1
            if (pending.isEmpty) {
              answer = value
              lastCompleted = true
            }
          }
          slot += 1
        }
      }
      answer
    }
  }

  /** The buffered online controller with the service derived from its quantum.
    *
    * @param galil the wrapped controller (`source`, with its concrete type)
    */
  final class RealtimeGalil private (val galil: OnlineGalil) extends BufferedSource(galil, galil.timing.service) {
    def acceptsEmpty: Boolean = RealtimeGalil.acceptsEmpty
  }

  object RealtimeGalil {
    def apply(quantum: Int = FPP_QUANTUM): RealtimeGalil = new RealtimeGalil(new OnlineGalil(quantum))

    /** Python `accepts_empty` (a static method): the empty word is a palindrome. */
    def acceptsEmpty: Boolean = true
  }
}

File: docs/palindromes-in-peg/galil_realtime.py
"""Executable FIFO service specification; local SCA lowering is separate.

The deque models the buffer for this reference. It is not claimed to be a
finite local transition table. Output decisions depend only on source events
and buffer emptiness; diagnostic counts do not guide the source or scheduler.
"""
from collections import deque

from scaffold_galil import OnlineGalil, FPP_QUANTUM


class BufferedSource:
  def __init__(self, source, service):
    if type(service) is not int or service < 1:
      raise ValueError("positive integral source service required")
    self.source, self.service, self.pending = source, service, deque()
    self.offered = self.reads = self.outputs = self.transitions = 0
    self.max_pending = 0
    self.last_completed = False

  def read(self, char):
    if char not in ("a", "b"):
      raise ValueError("one binary input symbol required")
    self.pending.append(char)
    self.offered += 1
    self.max_pending = max(self.max_pending, len(self.pending))
    answer, self.last_completed = 0, False
    for _ in range(self.service):
      if self.source.input_ready:
        if not self.pending:
          break  # remaining service transitions are quiescent
        result = self.source.read(self.pending.popleft())
        self.reads += 1
      else:
        result = self.source.work()
      self.transitions += 1
      if result.output is not None:
        self.outputs += 1
        if not self.pending:
          answer, self.last_completed = result.output, True
    return answer


class RealtimeGalil(BufferedSource):
  def __init__(self, quantum=FPP_QUANTUM):
    source = OnlineGalil(quantum)
    super().__init__(source, source.timing.service)

  @staticmethod
  def accepts_empty():
    return True
